pytorch_lightning/loops/optimizer/optimizer_loop.py

```diff
@@ -0,0 +1,376 @@
+
+    def __init__(self):
+        super().__init__()
+        # TODO: use default dict here to simplify logic in loop
+        self.outputs: _OUTPUTS_TYPE = []
+        self.optim_progress: OptimizationProgress = OptimizationProgress()
+
+        self._skip_backward: bool = False
+        self._batch_idx: Optional[int] = None
+        self._optimizers: Optional[List[Optimizer]] = None
+        self._hiddens: Optional[Any] = None
+
+    @property
+    def done(self) -> bool:
+        """Returns ``True`` when the last optimizer in the sequence has run."""
+        return self.optim_progress.optimizer_idx >= len(self._optimizers)
+
+    def connect(self, **kwargs: "Loop") -> None:
+        raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")
+
+    def reset(self) -> None:
+        if not self.restarting:
+            self.optim_progress.optimizer_idx = 0
+        self.outputs = [[] for _ in range(len(self.trainer.optimizers))]
+
+    def on_run_start(self, batch: Any, hiddens: Any, optimizers: List[Optimizer], batch_idx: int) -> None:
+        self._batch_idx = batch_idx
+        self._optimizers = optimizers.copy()
+
+    def advance(self, batch: Any, hiddens: Any, *args, **kwargs) -> None:
+        self._hiddens = hiddens
+        result = self._run_optimization(
+            self._batch_idx,
+            batch,
+            self.optim_progress.optimizer_idx,
+            self._optimizers[self.optim_progress.optimizer_idx],
+        )
+        if result:
+            self.outputs[self.optim_progress.optimizer_idx].append(copy(result.result_collection))
+
+        self.optim_progress.optimizer_idx += 1
+
+    def on_run_end(self) -> Tuple[_OUTPUTS_TYPE, Optional[Any]]:
+        outputs = self.outputs
+        hiddens = self._hiddens
+        # free memory
+        self.outputs = []
+        self._hiddens = None
+        return outputs, hiddens
+
+    def backward(
+        self,
+        loss: Tensor,
+        optimizer: Optional[torch.optim.Optimizer],
+        opt_idx: Optional[int] = None,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Tensor:
+        """Performs the backward step.
+
+        Args:
+            loss: The loss value to back-propagate on
+            optimizer: Current optimizer being used. ``None`` if using manual optimization.
+            opt_idx: Index of the current optimizer being used. ``None`` if using manual optimization.
+        """
+        self.trainer.accelerator.backward(loss, optimizer, opt_idx, *args, **kwargs)
+
+        if not self.trainer.fit_loop.should_accumulate():
+            # track gradients
+            grad_norm_dict = self._track_and_norm_grad(optimizer=optimizer)
+            if grad_norm_dict:
+                self.trainer.lightning_module._current_fx_name = "on_after_backward"
+                self.trainer.lightning_module.log_grad_norm(grad_norm_dict)
+        return loss
+
+    def _run_optimization(
```

Comment:
Do you think this could also become a utility ?

```py
def run_optimization(
    closure: LightningClosure,
    should_accumulate,
    optimizer,
    opt_idx,
    trainer,
    optimizer_step,
) -> Optional[ClosureResult]:
    """Runs closure (train step + backward) together with optimization if necessary.
    Args:
        batch_idx: the index of the current batch
        split_batch: the current tbptt split of the whole batch
        opt_idx: the index of the current optimizer or `None` in case of manual optimization
        optimizer: the current optimizer or `None` in case of manual optimization
    """
    run_optimization_start(opt_idx, optimizer)

    if should_accumulate:
        # For gradient accumulation

        # -------------------
        # calculate loss (train step + train step end)
        # -------------------
        # automatic_optimization=True: perform ddp sync only when performing optimizer_step
        with _block_parallel_sync_behavior(trainer, block=True):
            closure()

    # ------------------------------
    # BACKWARD PASS
    # ------------------------------
    # gradient update with accumulated gradients
    else:
        optimizer_step(optimizer, opt_idx, batch_idx, closure)

    result = closure.get_result()

    # untoggle model params
    run_optimization_end(opt_idx)
    return result
```

Reply:
you are right, I think that is useful!

Reply:
added this as a direct follow up #7938 as it is easier to do once this PR is merged